cognite/powerops/resync/v2/shop_to_assets.py

```diff
@@ -0,0 +1,370 @@
+            penalty_limit=data.get("shop_penalty_limit", self.default_shop_penalty_limit),
+            plants=plants,
+        )
+        if watercourse.external_id in existing:
+            raise ValueError(f"Watercourse with external id {watercourse.external_id} already exists")
+        assets_by_xid[watercourse.external_id] = watercourse
+
+        for price_area in watercourse_config.get("market_to_price_area", {}).values():
+            if price_area in existing:
+                raise ValueError(f"Price area with external id {price_area} already exists")
+            price_area = PriceAreaWrite(
+                external_id=f"price_area_{price_area}", name=price_area, timezone=self.default_timezone
+            )
+            assets_by_xid[price_area.external_id] = price_area
+
+        return assets_by_xid
+
+    def _to_generator(self, name: str, data: dict) -> GeneratorWrite:
+        curve = data["gen_eff_curve"]
+        efficiency_curve = GeneratorEfficiencyCurveWrite(
+            external_id=f"{name}_efficiency_curve", ref=curve["ref"], power=curve["x"], efficiency=curve["y"]
+        )
+        turbine_curves = []
+        for curve in data["turb_eff_curves"]:
+            turbine_curve = TurbineEfficiencyCurveWrite(
+                external_id=f"{name}_turbine_eff_curve_ref_{curve['ref']}",
+                head=curve["ref"],
+                flow=curve["x"],
+                efficiency=curve["y"],
+            )
+            turbine_curves.append(turbine_curve)
+
+        # Start cose is assumed to be a dictionary with (timestamp, value) key-value pairs
+        # We assume it is constant and thus use only the first value.
+        startcost = next(iter(data["startcost"].values()))
+
+        generator_timeseries = self.times_series_by_generator_name.get(name, {})
+
+        return GeneratorWrite(
+            external_id=f"generator_{name}",
+            name=name,
+            p_min=data["p_min"],
+            penstock=data["penstock"],
+            start_cost=startcost,
+            start_stop_cost=generator_timeseries.get("start_stop_cost"),
+            is_available_time_series=generator_timeseries.get("is_available"),
+            efficiency_curve=efficiency_curve,
+            turbine_curves=turbine_curves,
+        )
+
+    def _to_plant(
+        self,
+        name: str,
+        data: dict,
+        plant_display_name_and_order: dict,
+        watercourse_xid: str,
+        reservoir_by_name: dict[str, ReservoirWrite],
+        generator_by_name: dict[str, GeneratorWrite],
+        shop_model: dict,
+    ) -> PlantWrite:
+        plant_timeseries = self.times_series_by_plant_name.get(name, {})
+        display_name, order = plant_display_name_and_order.get(name, (name, 999))
+
+        all_connections = shop_model["connections"]
+        all_junctions = shop_model["model"].get("junction", {})
+        all_tunnels = shop_model["model"].get("tunnel", {})
```

Comment:
**suggestion:** add type hints since an assumption is being made on what these objects look like